lib/runners/workspace/git.rb

```diff
@@ -0,0 +1,49 @@
+module Runners
+  class Workspace::Git < Workspace
+    private
+
+    def prepare_base_source(dest)
+      base = git_source.base
+      if base
+        provision(base, dest)
+      else
+        raise ArgumentError, "base must not be nil"
+      end
+    end
+
+    def prepare_head_source(dest)
+      provision(git_source.head, dest)
+    end
+
+    def provision(commit_hash, dest)
+      shell = Shell.new(current_dir: git_directory, trace_writer: trace_writer, env_hash: {})
+      shell.capture3!("git checkout #{commit_hash}")
+      FileUtils.copy_entry(git_directory, dest)
+      FileUtils.remove_entry(dest / ".git")
+    end
+
+    def git_directory
+      @git_directory ||= root_tmp_dir.tap do |path|
+        shell = Shell.new(current_dir: path, trace_writer: trace_writer, env_hash: {})
+        shell.capture3!("git init")
+        shell.capture3!("git config gc.auto 0")
+        shell.capture3!("git remote add origin #{remote_url}")
```

Comment:
[question] Is there no problems when executing in shell mode?

```suggestion
        shell.capture3!("git", "remote", "add", "origin", remote_url)
```

Reply:
I don't know, but I will adjust to other implementations.